syntax = "proto3";

package cursor;

// Main cursor message
message CursorMessage {
    MessageType type = 1;

    oneof payload {
        CursorData cursor_data = 2;      // Cursor image data (sent once per cursor, cached by client)
        CursorSignal cursor_signal = 3;  // Cursor switch signal (sent on change)
        NativeCursor native_cursor = 5;  // Use client's native/local cursor rendering
    }

    uint64 timestamp = 4;
}

// Cursor image data - WebP encoded, sent once per unique cursor
message CursorData {
    string cursor_id = 1;           // Unique ID (hash of cursor image)
    bytes image_data = 2;           // WebP encoded cursor image
    int32 width = 3;                // Original image width in pixels
    int32 height = 4;               // Original image height in pixels
    int32 hotspot_x = 5;            // Hotspot X in pixels (relative to image)
    int32 hotspot_y = 6;            // Hotspot Y in pixels (relative to image)
    float dpi_scale = 7;            // Server-side DPI scale factor (e.g. 1.0, 1.25, 1.5, 2.0)
    bool is_animated = 8;           // True if this is a frame of an animated cursor
    uint32 frame_delay_ms = 9;      // Estimated delay between animation frames (ms)
}

// Cursor switch signal - lightweight, tells client which cached cursor to show
message CursorSignal {
    string cursor_id = 1;           // Reference to cached cursor image
}

// Native cursor signal - tells client to use local/native cursor rendering
// Used for XOR/inversion cursors that cannot be accurately captured as images
message NativeCursor {
    string cursor_name = 1;         // CSS cursor value: "default", "text", "crosshair", "move", etc.
}

enum MessageType {
    MESSAGE_TYPE_UNSPECIFIED = 0;
    MESSAGE_TYPE_CURSOR_DATA = 1;    // New cursor image (client should cache)
    MESSAGE_TYPE_CURSOR_SIGNAL = 2;  // Switch to cached cursor
    MESSAGE_TYPE_CURSOR_HIDE = 3;    // Cursor hidden
    MESSAGE_TYPE_HEARTBEAT = 4;      // Heartbeat
    MESSAGE_TYPE_CURSOR_NATIVE = 5;  // Use client's native cursor
}



